File: financial-app/lib/math-engine.ts
// Financial Math Engine — ported from Python math_engine.py
// TODO: move to shared — canonical algorithm spec in shared/README.md
// Cross-language duplicate: financial-tool/models/math_engine.py

// ── Basic TVM ──────────────────────────────────────────────────────────────

export function fv(pv: number, annualRate: number, years: number): number {
  return pv * Math.pow(1 + annualRate, years)
}

export function pvFromFv(fvVal: number, annualRate: number, years: number): number {
  return fvVal / Math.pow(1 + annualRate, years)
}

export function cagr(pvVal: number, fvVal: number, years: number): number {
  if (pvVal <= 0 || years <= 0) return 0
  return Math.pow(fvVal / pvVal, 1 / years) - 1
}

// ── Annuity ────────────────────────────────────────────────────────────────

export function annuityFv(payment: number, periodicRate: number, periods: number): number {
  if (periodicRate === 0) return payment * periods
  return payment * (Math.pow(1 + periodicRate, periods) - 1) / periodicRate
}

export function annuityPv(payment: number, periodicRate: number, periods: number): number {
  if (periodicRate === 0) return payment * periods
  return payment * (1 - Math.pow(1 + periodicRate, -periods)) / periodicRate
}

// ── Retirement Reverse Solver ──────────────────────────────────────────────

export function monthlySavingsNeeded(
  targetFv: number,
  currentSavings: number,
  annualRate: number,
  months: number,
): number {
  const rm = annualRate / 12
  if (months <= 0) return 0
  if (rm === 0) return Math.max(0, (targetFv - currentSavings) / months)

  const growthFactor = Math.pow(1 + rm, months)
  const shortfall = targetFv - currentSavings * growthFactor
  if (shortfall <= 0) return 0

  return shortfall / ((growthFactor - 1) / rm)
}

export function yearsToTarget(
  targetFv: number,
  currentSavings: number,
  monthlyContribution: number,
  annualRate: number,
  maxYears = 100,
): number {
  const rm = annualRate / 12
  const balanceAt = (m: number) =>
    currentSavings * Math.pow(1 + rm, m) + annuityFv(monthlyContribution, rm, m)

  if (balanceAt(maxYears * 12) < targetFv) return Infinity

  let lo = 0, hi = maxYears * 12
  for (let i = 0; i < 60; i++) {
    const mid = (lo + hi) / 2
    balanceAt(Math.floor(mid)) >= targetFv ? (hi = mid) : (lo = mid)
  }
  return hi / 12
}

// ── Compound Growth Series ─────────────────────────────────────────────────

export interface CompoundSeries {
  years: number[]
  balance: number[]
  contributions: number[]
  interest: number[]
}

export function compoundSeries(
  initial: number,
  annualRate: number,
  years: number,
  monthlyContribution = 0,
  compoundFreq = 12,
): CompoundSeries {
  const rp = annualRate / compoundFreq
  const contribsPerPeriod = (monthlyContribution * 12) / compoundFreq
  const totalPeriods = years * compoundFreq

  const yearLabels: number[] = []
  const balances: number[] = []
  const contributions: number[] = []
  const interests: number[] = []

  let balance = initial
  let totalContrib = initial
  let totalInterest = 0

  for (let period = 0; period <= totalPeriods; period++) {
    yearLabels.push(Math.round((period / compoundFreq) * 10000) / 10000)
    balances.push(Math.round(balance * 100) / 100)
    contributions.push(Math.round(totalContrib * 100) / 100)
    interests.push(Math.round(totalInterest * 100) / 100)

    if (period < totalPeriods) {
      const earned = balance * rp
      balance += earned + contribsPerPeriod
      totalContrib += contribsPerPeriod
      totalInterest += earned
    }
  }

  return { years: yearLabels, balance: balances, contributions, interest: interests }
}

// ── IRR ───────────────────────────────────────────────────────────────────

export function irr(cashFlows: number[], maxIter = 1000, tol = 1e-7): number {
  let rate = 0.1
  for (let i = 0; i < maxIter; i++) {
    const npvVal = cashFlows.reduce((s, cf, t) => s + cf / Math.pow(1 + rate, t), 0)
    const dNpv = cashFlows.reduce((s, cf, t) => s - (t * cf) / Math.pow(1 + rate, t + 1), 0)
    if (Math.abs(dNpv) < 1e-12) break
    const newRate = rate - npvVal / dNpv
    if (Math.abs(newRate - rate) < tol) return newRate
    rate = Math.max(-0.999, newRate)
  }
  return rate
}

export function npv(cashFlows: number[], discountRate: number): number {
  return cashFlows.reduce((s, cf, t) => s + cf / Math.pow(1 + discountRate, t), 0)
}

File: financial-tool/models/math_engine.py
"""
Financial Math Engine
FV / PV / IRR / CAGR / Annuity / Retirement reverse solver

TODO: move to shared — canonical algorithm spec in shared/README.md
Cross-language duplicate: financial-app/lib/math-engine.ts
"""

from typing import List


# ── Basic TVM ──────────────────────────────────────────────────────────────

def fv(pv: float, annual_rate: float, years: float) -> float:
    """Future Value: FV = PV * (1 + r)^n"""
    return pv * (1 + annual_rate) ** years


def pv_from_fv(fv_val: float, annual_rate: float, years: float) -> float:
    """Present Value: PV = FV / (1 + r)^n"""
    if (1 + annual_rate) ** years == 0:
        return 0
    return fv_val / (1 + annual_rate) ** years


def cagr(pv_val: float, fv_val: float, years: float) -> float:
    """CAGR = (FV / PV)^(1/n) - 1"""
    if pv_val <= 0 or years <= 0:
        return 0.0
    return (fv_val / pv_val) ** (1.0 / years) - 1


# ── Annuity ────────────────────────────────────────────────────────────────

def annuity_fv(payment: float, periodic_rate: float, periods: int) -> float:
    """
    Annuity Future Value (end of period payments).
    FV = P * [(1+r)^n - 1] / r
    """
    if periodic_rate == 0:
        return payment * periods
    return payment * ((1 + periodic_rate) ** periods - 1) / periodic_rate


def annuity_pv(payment: float, periodic_rate: float, periods: int) -> float:
    """
    Annuity Present Value (end of period payments).
    PV = P * [1 - (1+r)^-n] / r
    """
    if periodic_rate == 0:
        return payment * periods
    return payment * (1 - (1 + periodic_rate) ** (-periods)) / periodic_rate


# ── Retirement Reverse Solver ──────────────────────────────────────────────

def monthly_savings_needed(
    target_fv: float,
    current_savings: float,
    annual_rate: float,
    months: int,
) -> float:
    """
    How much to save per month to reach target_fv?

    target_fv = current_savings*(1+r_m)^n + PMT*[(1+r_m)^n - 1]/r_m
    → PMT = [target_fv - current_savings*(1+r_m)^n] * r_m / [(1+r_m)^n - 1]
    """
    r_m = annual_rate / 12
    if months <= 0:
        return 0.0
    if r_m == 0:
        remaining = target_fv - current_savings
        return max(0.0, remaining / months)

    growth_factor = (1 + r_m) ** months
    existing_growth = current_savings * growth_factor
    shortfall = target_fv - existing_growth

    if shortfall <= 0:
        return 0.0

    annuity_factor = (growth_factor - 1) / r_m
    return shortfall / annuity_factor


def years_to_target(
    target_fv: float,
    current_savings: float,
    monthly_contribution: float,
    annual_rate: float,
    max_years: int = 100,
) -> float:
    """Binary search: how many years until savings hit target?"""
    r_m = annual_rate / 12

    def balance_at(months: int) -> float:
        b = current_savings * (1 + r_m) ** months
        b += annuity_fv(monthly_contribution, r_m, months)
        return b

    if balance_at(max_years * 12) < target_fv:
        return float("inf")

    lo, hi = 0, max_years * 12
    for _ in range(60):
        mid = (lo + hi) / 2
        if balance_at(int(mid)) >= target_fv:
            hi = mid
        else:
            lo = mid
    return hi / 12


# ── Compound Growth Time Series ────────────────────────────────────────────

def compound_series(
    initial: float,
    annual_rate: float,
    years: int,
    monthly_contribution: float = 0.0,
    compound_freq: int = 12,  # 1=annual, 4=quarterly, 12=monthly, 365=daily
) -> dict:
    """
    Generate period-by-period balance breakdown.
    Returns dict with lists: years, balance, contributions, interest
    """
    r_p = annual_rate / compound_freq
    contribs_per_period = monthly_contribution * 12 / compound_freq
    total_periods = years * compound_freq

    year_labels, balances, contributions, interests = [], [], [], []

    balance = initial
    total_contrib = initial
    total_interest = 0.0

    for period in range(total_periods + 1):
        year_labels.append(round(period / compound_freq, 4))
        balances.append(round(balance, 2))
        contributions.append(round(total_contrib, 2))
        interests.append(round(total_interest, 2))

        if period < total_periods:
            earned = balance * r_p
            balance += earned + contribs_per_period
            total_contrib += contribs_per_period
            total_interest += earned

    return {
        "years": year_labels,
        "balance": balances,
        "contributions": contributions,
        "interest": interests,
    }


# ── IRR (Newton-Raphson) ───────────────────────────────────────────────────

def irr(cash_flows: List[float], max_iter: int = 1000, tol: float = 1e-7) -> float:
    """
    Internal Rate of Return via Newton-Raphson.
    cash_flows[0] is typically negative (initial outflow).
    """
    rate = 0.1
    for _ in range(max_iter):
        npv_val = sum(cf / (1 + rate) ** t for t, cf in enumerate(cash_flows))
        d_npv = sum(
            -t * cf / (1 + rate) ** (t + 1) for t, cf in enumerate(cash_flows)
        )
        if abs(d_npv) < 1e-12:
            break
        new_rate = rate - npv_val / d_npv
        if abs(new_rate - rate) < tol:
            return new_rate
        rate = max(-0.999, new_rate)  # prevent divide-by-zero
    return rate


def npv(cash_flows: List[float], discount_rate: float) -> float:
    """Net Present Value"""
    return sum(cf / (1 + discount_rate) ** t for t, cf in enumerate(cash_flows))
